ts/client/storageEngine/transactions/db.ts

```diff
@@ -0,0 +1,197 @@
+import { txDB } from "../../database/connection";
+import { OffchainTx } from "../../features/interface";
+import { TransferOffchainTx } from "../../features/transfer";
+import { Status } from "./constants";
+import {
+    SubmitMeta,
+    SyncMeta,
+    TransactionStatus,
+    TransactionStorage,
+    TransationMessageOrObject
+} from "./interfaces";
+
+/**
+ * levelDB implementation of TransactionStorage
+ */
+export class TransactionDBStorage implements TransactionStorage {
+    private readonly db = txDB;
+
+    public async get(
+        msgOrTxn: TransationMessageOrObject
+    ): Promise<TransactionStatus> {
+        return this.fromDB(this.getMessage(msgOrTxn));
+    }
+
+    public async pending(txn: OffchainTx): Promise<TransactionStatus> {
+        const txnMessage = txn.message();
+        await this.throwIfAlreadyExists(txnMessage);
+
+        const txnStatus = {
+            status: Status.Pending,
+            transaction: txn
+        };
+
+        await this.toDB(txnMessage, txnStatus);
+        return txnStatus;
+    }
+
+    public async submitted(
+        msgOrTxn: TransationMessageOrObject,
+        { batchID, l1TxnHash, l1BlockIncluded }: SubmitMeta
+    ): Promise<TransactionStatus> {
+        const msg = this.getMessage(msgOrTxn);
+        const txnStatus = await this.getStatusOrFail(msg);
+        this.validateStatusTransition(txnStatus.status, Status.Submitted);
+        const submittedTxnStatus = {
+            ...txnStatus,
+            status: Status.Submitted,
+            batchID,
+            l1TxnHash,
+            l1BlockIncluded
+        };
+        await this.toDB(msg, submittedTxnStatus);
+        return submittedTxnStatus;
+    }
+
+    public async finalized(
+        msgOrTxn: TransationMessageOrObject
+    ): Promise<TransactionStatus> {
+        return this.transition(this.getMessage(msgOrTxn), Status.Finalized);
+    }
+
+    public async failed(
+        msgOrTxn: TransationMessageOrObject,
+        detail: string
+    ): Promise<TransactionStatus> {
+        return this.transition(
+            this.getMessage(msgOrTxn),
+            Status.Failed,
+            detail
+        );
+    }
+
+    public async sync(
+        txn: OffchainTx,
+        { batchID, l1TxnHash, l1BlockIncluded, finalized }: SyncMeta
+    ): Promise<TransactionStatus> {
+        const txnMessage = txn.message();
+        await this.throwIfAlreadyExists(txnMessage);
+
+        const status = finalized ? Status.Finalized : Status.Submitted;
+        const txnStatus = {
+            transaction: txn,
+            status,
+            batchID,
+            l1TxnHash,
+            l1BlockIncluded
+        };
+        await this.toDB(txnMessage, txnStatus);
+        return txnStatus;
+    }
+
+    public async count(): Promise<number> {
+        const stream = this.db.createKeyStream();
+        let count = 0;
+        for await (const _ of stream) {
+            count++;
+        }
+        return count;
```

Comment:
Another way we could do this is to store the count at a static key. This would require us to update the count on every insert (increased write time), but would make reads very quick. And in the worst case on a key miss we would have to do this count anyways. Would look like:

1. Check for count key. If exists return value.
2. If value missing at key, run count from stream. Write result and return.

In addition to needing to increment count when inserting new values. Think we can skip this optimization for now and consider it in the future if this function becomes a bottleneck.

Reply:
I did explore this when I was trying to save Batches in the DB (upcoming PR). There is no inherent issue with this (infact it has some advantages like you mentioned) but saving `count` in a db structure with `key = msg` and `value = serialized tx` felt a little unintuitive to me as the db was not "meant to" store that information. Also something which is enforced in the SQL land, keeping a db object pure.

Open to interpretation though.

Reply:
We could always add a new 'table' (subdb) that is just for storing counts with the key being related to the entity it is storing if we wanted to have that explicit separation. I don't think there is a ton of harm in a k/v store of having different values being stored in the same area, as the key delimiters can be used to separate types of data.

For example, if we had a Cat area (subdb) and we also wanted to include a counter cache for it, we might key it like:
```
cat_123: ... // cat 1 serialized data
cat_456: ... // cat 2 serialized data
cat_num_cats: 2
``` 
And then having a TS const like `NUM_CATS_KEY = "num_cats"`. It would be very clear in code when writes/reads are modifying the count and not a kitty cat.

All this being said, don't think we need this level of per-optimzation yet and we could come back to it. Feel free to resolve.

Reply:
Right makes sense. Thanks 😄 
Just need to make note of not including the cache/metadata per db when doing things like counting keys etc as `createStream` will include these keys as well.